File: solutions/go/src/playground/algorithm/reverse_linked_list.go
package algorithm

func reverseList(head *ListNode) *ListNode {
	return reverseListR(nil, head)
}

func reverseListR(previous *ListNode, current *ListNode) *ListNode {
	if current == nil {
		return previous
	}
	head := reverseListR(current, current.Next)
	current.Next = previous
	return head
}

File: solutions/java/src/main/java/playground/algorithm/ReverseLinkedList.java
package playground.algorithm;

import playground.algorithm.common.ListNode;

public class ReverseLinkedList {

    public ListNode reverseList(ListNode head) {
        ListNode last = null;
        ListNode current = head;
        ListNode next;
        while (current != null) {
            next = current.next;
            current.next = last;
            last = current;
            current = next;
        }
        return last;
    }

    public ListNode reverseListRecursively(ListNode head) {
        return reverseListRecursively(null, head);
    }

    private ListNode reverseListRecursively(ListNode current, ListNode next) {
        if (next == null) {
            return current;
        }
        ListNode head = reverseListRecursively(next, next.next);
        next.next = current;
        return head;
    }
}
